### DISCUSSION TITLE
Why Doesn't FastAPI Support Dependency Injection in the Startup Process?

### DISCUSSION THREAD

**Discussion by Johnny850807** [fastapi/fastapi#12082] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.on_event("startup")
async def startup_event():
                   repo=Depends(init_my_repository), 
                   xxx=Depends(init_my_XXX), 
                   yyy=Depends(init_my_YYY)):
    repo.do_something()
    xxx.do_something()
    yyy.do_something()
```

## Why Doesn't FastAPI Support Dependency Injection in the Startup Process?

I'm facing a frustrating limitation with FastAPI's dependency injection system, and I need some help understanding why this isn't supported more naturally.

In my application, I have several complex dependencies, a simple case is that -- a `Repository` class that relies on a MongoDB database connection. These dependencies require significant setup, and I rely on FastAPI's `Depends` mechanism to handle this. 

However, when I need to perform some initialization tasks during the application's startup, I hit a roadblock.

For example, I want to do something like this:

```python
@asynccontextmanager
async def lifespan(app: FastAPI, 
                   repo=Depends(init_my_repository), 
                   xxx=Depends(init_my_XXX), 
                   yyy=Depends(init_my_YYY)):
    repo.do_something()
    xxx.do_something()
    yyy.do_something()
    yield
```

or
```python
@app.on_event("startup")
async def startup_event():
                   repo=Depends(init_my_repository), 
                   xxx=Depends(init_my_XXX), 
                   yyy=Depends(init_my_YYY)):
    repo.do_something()
    xxx.do_something()
    yyy.do_something()
    ....
```

But FastAPI doesn't support injecting dependencies this way. 

Many people suggest that I create a singleton instance myself, but the point is, what if my class has a very complex dependency structure? It might depend on a whole bunch of sub-dependencies and even sub-sub-dependencies. If I have to manually create these dependencies instead of relying on IoC to handle them, it feels incredibly inefficient and counterproductive. Or am I missing something here?

Given how common this scenario is in other IoC-based frameworks, it's disappointing that FastAPI doesn't offer a straightforward solution. I constantly feel like I'm hacking my application and I have spent 3 days on this simple task..... (very painful)

Although I've seen other discussions on this topic, I still want to push a bit further and ask if there are any plans to include this feature in the roadmap. Everything has a reason, so I'd like to understand what limitations are preventing FastAPI from supporting such a fundamental functionality.

Thank you very much!

### Operating System

macOS

### FastAPI Version

0.112.1

### Pydantic Version

2.8.2

### Python Version

Python 3.11.9

**Comment by YuriiMotov** [comment, marked as answer]
Have you seen this: https://github.com/fastapi/fastapi/discussions/11742 ?

You can either use the approach mentioned above or use external dependency injection system
